File: assets/js/engine/report.js
/**
 * Diagnóstico de horas no colocadas — puerto JS de `backend/app/solver/report.py`.
 *
 * "No se pudo" no le sirve a una escuela. Para cada (grupo, materia) con horas
 * faltantes se determina la causa dominante y se redacta la acción que la destraba.
 *
 * @module engine/report
 */

import { groupLabel } from './contract.js';

export function diagnoseUnplaced(ctx, assignment, unplacedUids) {
  if (!unplacedUids.length) return [];

  const lessons = new Map(ctx.lessons.map((l) => [l.uid, l]));
  const grouped = new Map();
  for (const uid of unplacedUids) {
    const lesson = lessons.get(uid);
    const key = `${lesson.groupId}|${lesson.subjectId}`;
    grouped.set(key, (grouped.get(key) || 0) + 1);
  }

  const groupBusy = new Map();
  const teacherBusy = new Map();
  const teacherHours = new Map();
  for (const [uid, { slot, teacherId }] of assignment) {
    const lesson = lessons.get(uid);
    if (!groupBusy.has(lesson.groupId)) groupBusy.set(lesson.groupId, new Set());
    groupBusy.get(lesson.groupId).add(slot);
    if (!teacherBusy.has(teacherId)) teacherBusy.set(teacherId, new Set());
    teacherBusy.get(teacherId).add(slot);
    teacherHours.set(teacherId, (teacherHours.get(teacherId) || 0) + 1);
  }

  const out = [];
  for (const [key, missing] of [...grouped.entries()].sort()) {
    const [gid, sid] = key.split('|');
    const group = ctx.groups.get(gid);
    const subject = ctx.subjects.get(sid);
    const reason = explain(ctx, gid, sid, groupBusy.get(gid) || new Set(), teacherBusy, teacherHours);
    out.push({
      severity: 'error',
      code: 'UNPLACED_HOURS',
      message: `Faltaron ${missing} h de ${subject?.name || sid} en ${groupLabel(group)}. ${reason}`,
      group_id: gid,
      subject_id: sid,
      teacher_id: null,
      missing_hours: missing,
    });
  }
  return out;
}

function explain(ctx, gid, sid, groupBusy, teacherBusy, teacherHours) {
  const eligible = ctx.eligible.get(`${gid}|${sid}`) || [];
  if (!eligible.length) {
    return 'Ningún profesor puede impartirla: asígnala a alguien de la plantilla.';
  }

  const freeGroup = new Set([...ctx.groupSlots.get(gid)].filter((s) => !groupBusy.has(s)));
  if (!freeGroup.size) {
    return 'El grupo ya no tiene espacios libres en su rejilla; su plan de estudios llena la semana.';
  }

  const saturated = [];
  const noOverlap = [];
  for (const tid of eligible) {
    const teacher = ctx.teachers.get(tid);
    if ((teacherHours.get(tid) || 0) >= ctx.teacherMaxWeekly.get(tid)) {
      saturated.push(teacher.name);
      continue;
    }
    const busy = teacherBusy.get(tid) || new Set();
    const tSlots = ctx.teacherSlots.get(tid);
    const overlap = [...freeGroup].some((s) => tSlots.has(s) && !busy.has(s));
    if (!overlap) noOverlap.push(teacher.name);
  }

  if (saturated.length && saturated.length === eligible.length) {
    return `Sus profesores (${saturated.join(', ')}) ya llegaron a su carga máxima semanal. ` +
      'Sube su tope de horas o reparte la materia con otro docente.';
  }
  if (noOverlap.length && noOverlap.length + saturated.length === eligible.length) {
    return `Las horas libres del grupo no coinciden con la disponibilidad de ${noOverlap.join(', ')} ` +
      '(o ya están dando clase a otro grupo en esos bloques). Libera disponibilidad o mueve otras materias.';
  }
  return 'Sí existen huecos sueltos, pero ninguno compatible al mismo tiempo con el grupo y con sus ' +
    "profesores. Suele resolverse ampliando la disponibilidad de un docente, subiendo 'máx. por día' " +
    'de la materia, o dando más segundos de cálculo.';
}

File: backend/app/solver/report.py
"""
Diagnóstico de las horas que no se pudieron colocar.

Devolver "no se pudo" no sirve a una escuela. Para cada (grupo, materia) con horas
faltantes se determina la CAUSA dominante inspeccionando el estado final, y se
redacta en el idioma del director, con la acción concreta que lo destraba.
"""

from __future__ import annotations

from collections import defaultdict
from typing import Dict, List, Tuple

from ..models.schemas import Conflict, Severity
from .domain import ProblemContext


def diagnose_unplaced(
    ctx: ProblemContext,
    assignment: Dict[str, Tuple[int, str]],
    unplaced_uids: List[str],
) -> List[Conflict]:
    if not unplaced_uids:
        return []

    lessons = {l.uid: l for l in ctx.lessons}
    grouped: Dict[Tuple[str, str], int] = defaultdict(int)
    for uid in unplaced_uids:
        lesson = lessons[uid]
        grouped[(lesson.group_id, lesson.subject_id)] += 1

    # Ocupación resultante, para poder explicar quién estorbó.
    group_busy: Dict[str, set] = defaultdict(set)
    teacher_busy: Dict[str, set] = defaultdict(set)
    teacher_hours: Dict[str, int] = defaultdict(int)
    for uid, (slot, tid) in assignment.items():
        group_busy[lessons[uid].group_id].add(slot)
        teacher_busy[tid].add(slot)
        teacher_hours[tid] += 1

    conflicts: List[Conflict] = []
    for (gid, sid), missing in sorted(grouped.items()):
        group = ctx.groups[gid]
        subject = ctx.subjects[sid]
        elig = ctx.eligible.get((gid, sid), tuple())
        reason = _explain(
            ctx, gid, sid, elig, group_busy[gid], teacher_busy, teacher_hours
        )
        conflicts.append(
            Conflict(
                severity=Severity.ERROR,
                code="UNPLACED_HOURS",
                message=(
                    f"Faltaron {missing} h de {subject.name} en {group.label}. {reason}"
                ),
                group_id=gid,
                subject_id=sid,
                missing_hours=missing,
            )
        )
    return conflicts


def _explain(
    ctx: ProblemContext,
    gid: str,
    sid: str,
    eligible: Tuple[str, ...],
    group_busy: set,
    teacher_busy: Dict[str, set],
    teacher_hours: Dict[str, int],
) -> str:
    if not eligible:
        return "Ningún profesor puede impartirla: asígnala a alguien de la plantilla."

    free_group = ctx.group_slots[gid] - group_busy
    if not free_group:
        return "El grupo ya no tiene espacios libres en su rejilla; su plan de estudios llena la semana."

    saturated: List[str] = []
    no_overlap: List[str] = []
    for tid in eligible:
        teacher = ctx.teachers[tid]
        if teacher_hours[tid] >= ctx.teacher_max_weekly[tid]:
            saturated.append(teacher.name)
            continue
        if not (ctx.teacher_slots[tid] & free_group) - teacher_busy[tid]:
            no_overlap.append(teacher.name)

    if saturated and len(saturated) == len(eligible):
        nombres = ", ".join(saturated)
        return (
            f"Sus profesores ({nombres}) ya llegaron a su carga máxima semanal. "
            "Sube su tope de horas o reparte la materia con otro docente."
        )
    if no_overlap and len(no_overlap) + len(saturated) == len(eligible):
        nombres = ", ".join(no_overlap)
        return (
            f"Las horas libres del grupo no coinciden con la disponibilidad de {nombres} "
            "(o ya están dando clase a otro grupo en esos bloques). "
            "Libera disponibilidad o mueve otras materias del grupo."
        )
    return (
        "Sí existen huecos sueltos, pero ninguno compatible al mismo tiempo con el grupo "
        "y con sus profesores. Suele resolverse ampliando la disponibilidad de un docente, "
        "subiendo 'max_per_day' de la materia, o dando más segundos de cálculo."
    )
